File: Matlab/run_small.m
% Load COIL20 dataset
%data_path = 'D:\0_Work\WSU\CS7900\Project\Rimon_Rojan_Adarsh\Rimon_Rojan_Adarsh\RUN\CS7900-project\Data\AR.mat', 'AR';
data_path = 'D:\0_Work\WSU\CS7900\Project\Rimon_Rojan_Adarsh\Rimon_Rojan_Adarsh\RUN\CS7900-project\Data\COIL20.mat', 'COIL20';
%data_path = 'D:\0_Work\WSU\CS7900\Project\Rimon_Rojan_Adarsh\Rimon_Rojan_Adarsh\RUN\CS7900-project\Data\MSRA25.mat', 'MSRA25';
%data_path = 'D:\0_Work\WSU\CS7900\Project\Rimon_Rojan_Adarsh\Rimon_Rojan_Adarsh\RUN\CS7900-project\Data\YaleB.mat', 'YaleB';


load(data_path);

% Take only first 5 classes, 10 samples per class
X = X';
Y = Y(:);
mask = Y <= 5;  % First 5 classes
X = X(:, mask);
Y = Y(mask);
X = X(:, 1:7:end);  % Take every 7th sample
Y = Y(1:7:end);

% Parameters
pca_dim = 15;
reduced_dim = 10;
h1 = 2;
h2 = 3;
r = 2;
alpha = 0.1;
maxiter = 5;

fprintf('MATLAB Results:\n');
fprintf('%s\n', repmat('-', 1, 50));
fprintf('Input X shape: %dx%d\n', size(X));
fprintf('Input Y shape: %dx%d\n', size(Y));
fprintf('Y unique values: '); fprintf('%d ', unique(Y)); fprintf('\n');
fprintf('Samples per class:\n');
tabulate(Y);

% PCA
meanX = mean(X, 2);
X_centered = X - meanX;
[U, ~, ~] = svd(X_centered, 'econ');
X_pca = U(:, 1:pca_dim)' * X_centered;

fprintf('\nAfter PCA:\n');
fprintf('Mean X first 3 values: '); fprintf('%.6f ', meanX(1:3)); fprintf('\n');
fprintf('X_pca shape: %dx%d\n', size(X_pca));
fprintf('X_pca first 3x3 values:\n');
disp(X_pca(1:3,1:3));

% Run ALLDA
[OBJ, W_allma, S1] = ALLDA(X_pca, Y, reduced_dim, h1, r, 1e-5);
Z_train_allma = W_allma' * X_pca;

fprintf('\nALLDA results:\n');
fprintf('W_allma shape: %dx%d\n', size(W_allma));
fprintf('W_allma first 3x3 values:\n');
disp(W_allma(1:3,1:3));
fprintf('Z_train_allma shape: %dx%d\n', size(Z_train_allma));
fprintf('Z_train_allma first 3x3 values:\n');
disp(Z_train_allma(1:3,1:3));

% Run ALLDA_semi
[W_semi, p, S, Obj] = ALLDA_semi(X_pca, Y, X_pca, h1, h2, reduced_dim, alpha, maxiter);
Z_train_semi = W_semi' * X_pca;

fprintf('\nALLDA_semi results:\n');
fprintf('W_semi shape: %dx%d\n', size(W_semi));
fprintf('W_semi first 3x3 values:\n');
disp(W_semi(1:3,1:3));
fprintf('Z_train_semi shape: %dx%d\n', size(Z_train_semi));
fprintf('Z_train_semi first 3x3 values:\n');
disp(Z_train_semi(1:3,1:3));


% Load Python results for comparison 
py_results = load('coil20_results.mat');
fprintf('\nComparison with Python results:\n');
fprintf('X_pca max difference: %.10f\n', max(abs(X_pca(:) - py_results.X_pca(:))));
fprintf('W_allma max difference: %.10f\n', max(abs(W_allma(:) - py_results.W_allma(:))));
fprintf('Z_train_allma max difference: %.10f\n', max(abs(Z_train_allma(:) - py_results.Z_train_allma(:))));
fprintf('W_semi max difference: %.10f\n', max(abs(W_semi(:) - py_results.W_semi(:))));
fprintf('Z_train_semi max difference: %.10f\n', max(abs(Z_train_semi(:) - py_results.Z_train_semi(:))));

File: Python/run_small.py
import numpy as np
from scipy.linalg import svd
from sklearn.neighbors import KNeighborsClassifier
import scipy.io as sio
from ALLDA import ALLDA
from ALLDA_semi import ALLDA_semi
from scipy.io import savemat

# Load COIL20 dataset
#data_path = r"D:\0_Work\WSU\CS7900\Project\Rimon_Rojan_Adarsh\Rimon_Rojan_Adarsh\RUN\CS7900-project\Data\AR.mat"
data_path = r"D:\0_Work\WSU\CS7900\Project\Rimon_Rojan_Adarsh\Rimon_Rojan_Adarsh\RUN\CS7900-project\Data\COIL20.mat"
#data_path = r"D:\0_Work\WSU\CS7900\Project\Rimon_Rojan_Adarsh\Rimon_Rojan_Adarsh\RUN\CS7900-project\Data\MSRA25.mat"
#data_path = r"D:\0_Work\WSU\CS7900\Project\Rimon_Rojan_Adarsh\Rimon_Rojan_Adarsh\RUN\CS7900-project\Data\YaleB.mat"


data = sio.loadmat(data_path)

# Take only first 5 classes, 10 samples per class
X = data['X'].T
Y = data['Y'].flatten().astype(int) - 1  # Convert to 0-based indexing
mask = Y < 5  # First 5 classes
X = X[:, mask]
Y = Y[mask]
X = X[:, ::7]  # Take every 7th sample to reduce size
Y = Y[::7]

# Parameters
pca_dim = 15
reduced_dim = 10
h1 = 2
h2 = 3
r = 2
alpha = 0.1
maxiter = 5

print("Python Results:")
print("-" * 50)
print(f"Input X shape: {X.shape}")
print(f"Input Y shape: {Y.shape}")
print(f"Y unique values: {np.unique(Y)}")
print(f"Samples per class: {np.bincount(Y)}")

# PCA
meanX = np.mean(X, axis=1, keepdims=True)
X_centered = X - meanX
U, s, Vt = svd(X_centered, full_matrices=False)
X_pca = (U[:, :pca_dim].T @ X_centered)

print(f"\nAfter PCA:")
print(f"Mean X first 3 values: {meanX.flatten()[:3]}")
print(f"X_pca shape: {X_pca.shape}")
print(f"X_pca first 3x3 values:\n{X_pca[:3,:3]}")

# Run ALLDA
OBJ, W_allma, S1 = ALLDA(X_pca, Y, reduced_dim, h1, r, 1e-5)
Z_train_allma = W_allma.T @ X_pca

print("\nALLDA results:")
print(f"W_allma shape: {W_allma.shape}")
print(f"W_allma first 3x3 values:\n{W_allma[:3,:3]}")
print(f"Z_train_allma shape: {Z_train_allma.shape}")
print(f"Z_train_allma first 3x3 values:\n{Z_train_allma[:3,:3]}")

# Run ALLDA_semi
W_semi, p, S, Obj = ALLDA_semi(X_pca, Y, X_pca, h1, h2, reduced_dim, alpha, maxiter)
Z_train_semi = W_semi.T @ X_pca

print("\nALLDA_semi results:")
print(f"W_semi shape: {W_semi.shape}")
print(f"W_semi first 3x3 values:\n{W_semi[:3,:3]}")
print(f"Z_train_semi shape: {Z_train_semi.shape}")
print(f"Z_train_semi first 3x3 values:\n{Z_train_semi[:3,:3]}")

matrices_to_save = {
    'X': X,
    'Y': Y,
    'meanX': meanX,
    'X_centered': X_centered,
    'U': U[:, :pca_dim],  # Save only used columns
    'X_pca': X_pca,
    'W_allma': W_allma,
    'S1': S1,
    'Z_train_allma': Z_train_allma,
    'W_semi': W_semi,
    'p': p,
    'S': S,
    'Z_train_semi': Z_train_semi,
    'OBJ_allda': OBJ,
    'Obj_semi': Obj
}

# Add comparison printing function
def print_matrix_info(name, matrix, num_elements=3):
    print(f"\n=== Matrix: {name} ===")
    print(f"Shape: {matrix.shape}")
    if isinstance(matrix, np.ndarray):
        print(f"First {num_elements} elements: ", end='')
        if matrix.size >= num_elements:
            print([f"{x:.6f}" for x in matrix.flatten()[:num_elements]])
        else:
            print([f"{x:.6f}" for x in matrix.flatten()])
        print(f"Min value: {np.min(matrix):.6f}")
        print(f"Max value: {np.max(matrix):.6f}")
        print(f"Mean value: {np.mean(matrix):.6f}")
    else:
        print(f"Type: {type(matrix)}")

# Save results and print detailed information
savemat('coil20_results.mat', matrices_to_save)

print("\nDetailed Matrix Information:")
print("=" * 50)

# Print information for key matrices
for name, matrix in matrices_to_save.items():
    if isinstance(matrix, np.ndarray):
        print_matrix_info(name, matrix)
    elif isinstance(matrix, dict):
        print(f"\n=== Dictionary: {name} ===")
        for k, v in matrix.items():
            if isinstance(v, np.ndarray):
                print_matrix_info(f"{name}.{k}", v)
    else:
        print(f"\n{name}: {type(matrix)}")

print("\nNote: These values can be compared with MATLAB output for verification.")